File: frontend/src/api/collections.ts
/**
 * Receivables, delivery and reminders.
 *
 * 🔴 Money is never formatted here. Every figure arrives with a `display`
 * string already grouped the Indian way, because the grouping rule lives in
 * `api/money.py` and a second implementation in TypeScript is a second one to
 * get wrong — the two would disagree by a rupee somewhere nobody looks.
 */

import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { api } from './client'

// ---------------------------------------------------------------------------
// Ageing
// ---------------------------------------------------------------------------

export interface AgeingRow {
  invoice_id: string
  invoice_no: string | null
  entity_code: string
  organisation_id: string | null
  buyer_name: string
  invoice_date: string
  due_date: string | null
  /** 🔴 True when there was no due date and 30 days was assumed. Say so. */
  due_date_assumed: boolean
  total_value: string
  amount_received: string
  amount_outstanding: string
  days_overdue: number
  bucket: string
  bucket_label: string
  status: string
  promised_on: string | null
  promise_note: string | null
  last_reminder_at: string | null
  reminder_count: number
  display: { total: string; outstanding: string }
}

export interface AgeingReport {
  summary: {
    as_of: string
    invoice_count: number
    total_outstanding: string
    assumed_due_dates: number
    buckets: Array<{
      bucket: string
      label: string
      count: number
      amount: string
      display: string
    }>
    display: { total_outstanding: string }
    note: string
  }
  rows: AgeingRow[]
  by_buyer: Array<{
    organisation_id: string | null
    buyer_name: string
    invoice_count: number
    total_outstanding: string
    oldest_days_overdue: number
    billing_opt_out: boolean
    billing_email: string | null
    display: { total_outstanding: string }
  }>
}

export function useAgeing(filters: Record<string, string> = {}) {
  const query = new URLSearchParams(
    Object.entries(filters).filter(([, v]) => v !== '' && v != null),
  ).toString()

  return useQuery({
    queryKey: ['receivables-ageing', filters],
    queryFn: () => api<AgeingReport>(`/api/v1/receivables/ageing/${query ? `?${query}` : ''}`),
  })
}

export interface PriorityFactor {
  factor: string
  value: unknown
  points: number
  explanation: string
}

export interface Priority {
  invoice_id: string
  score: number
  band: 'low' | 'medium' | 'high'
  /** 🔴 Render these. A score whose inputs are hidden cannot be argued with. */
  factors: PriorityFactor[]
  disclaimer: string
}

export function useCollectionPriority(limit = 25) {
  return useQuery({
    queryKey: ['receivables-priority', limit],
    queryFn: () => api<Priority[]>(`/api/v1/receivables/priority/?limit=${limit}`),
  })
}

// ---------------------------------------------------------------------------
// Payment requests and promises
// ---------------------------------------------------------------------------

export interface PaymentRequest {
  id: string
  invoice_id: string
  provider: string
  provider_reference: string | null
  amount: string
  currency: string
  payload_url: string | null
  qr_svg: string | null
  status: string
  expires_at: string | null
  created_at: string
  /** 🔴 Always false. Present so a screen cannot render this as a payment. */
  is_payment: boolean
  note: string
}

export function usePaymentRequests(invoiceId: string | undefined) {
  return useQuery({
    queryKey: ['payment-requests', invoiceId],
    queryFn: () => api<PaymentRequest[]>(`/api/v1/invoices/${invoiceId}/payment-requests/`),
    enabled: Boolean(invoiceId),
  })
}

export function useCreatePaymentRequest() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({
      invoiceId,
      ...body
    }: {
      invoiceId: string
      provider?: string
      amount?: string
      note?: string
      idempotency_key?: string
    }) =>
      api<PaymentRequest>(`/api/v1/invoices/${invoiceId}/payment-requests/`, {
        method: 'POST',
        body,
      }),
    onSuccess: (_, { invoiceId }) =>
      qc.invalidateQueries({ queryKey: ['payment-requests', invoiceId] }),
  })
}

export function useRecordPromise() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({
      invoiceId,
      ...body
    }: {
      invoiceId: string
      promised_on: string
      amount?: string
      note?: string
      contact_name?: string
    }) => api(`/api/v1/invoices/${invoiceId}/promises/`, { method: 'POST', body }),
    onSuccess: () => qc.invalidateQueries({ queryKey: ['receivables-ageing'] }),
  })
}

// ---------------------------------------------------------------------------
// Delivery
// ---------------------------------------------------------------------------

export interface DeliveryPreview {
  invoice_id: string
  channel: string
  recipient: string
  recipient_name: string | null
  subject: string | null
  body: string
  pdf_sha256: string | null
  template_version: string
  /** 🔴 Quote this back to send. Covers recipient, subject, body and the PDF. */
  preview_sha256: string
  warnings: string[]
  blocked_reason: string | null
  can_send: boolean
}

export interface Delivery {
  id: string
  invoice_id: string
  channel: string
  recipient: string
  status: string
  attempts: number
  pdf_sha256: string | null
  provider: string | null
  provider_message_id: string | null
  error_code: string | null
  error_detail: string | null
}

export function usePreviewDelivery() {
  return useMutation({
    mutationFn: ({
      invoiceId,
      ...body
    }: {
      invoiceId: string
      channel?: string
      recipient?: string
      subject?: string
      body?: string
      attach_pdf?: boolean
    }) =>
      api<DeliveryPreview>(`/api/v1/invoices/${invoiceId}/deliveries/preview/`, {
        method: 'POST',
        body,
      }),
  })
}

/**
 * 🔴 Requires the preview's hash. A send whose hash does not match the current
 * preview is refused — something changed between seeing it and confirming it.
 */
export function useSendDelivery() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({
      invoiceId,
      ...body
    }: {
      invoiceId: string
      preview_sha256: string
      channel?: string
      recipient?: string
      subject?: string
      body?: string
      attach_pdf?: boolean
    }) =>
      api<Delivery>(`/api/v1/invoices/${invoiceId}/deliveries/`, { method: 'POST', body }),
    onSuccess: (_, { invoiceId }) =>
      qc.invalidateQueries({ queryKey: ['deliveries', invoiceId] }),
  })
}

export interface DeliveryHistoryRow {
  id: string
  channel: string
  recipient: string
  recipient_name: string | null
  subject: string | null
  body_snapshot: string
  /** 🔴 The artifact this attempt carried, not whatever the invoice holds now. */
  pdf_sha256: string | null
  status: string
  attempts: number
  provider: string | null
  provider_message_id: string | null
  confirmed_at: string
  sent_at: string | null
  delivered_at: string | null
  failed_at: string | null
  error_code: string | null
  error_detail: string | null
  is_reminder: boolean
}

export function useDeliveryHistory(invoiceId: string | undefined) {
  return useQuery({
    queryKey: ['deliveries', invoiceId],
    queryFn: () => api<DeliveryHistoryRow[]>(`/api/v1/invoices/${invoiceId}/deliveries/`),
    enabled: Boolean(invoiceId),
  })
}

File: backend/schemas/collections.py
"""Shapes for receivables, payment requests, delivery and reminders."""

from __future__ import annotations

import uuid
from datetime import date
from decimal import Decimal
from typing import Any

from pydantic import BaseModel, Field

# ---------------------------------------------------------------------------
# Receivables
# ---------------------------------------------------------------------------


class AgeingBucket(BaseModel):
    bucket: str
    label: str
    count: int
    amount: str
    display: str


class AgeingSummary(BaseModel):
    as_of: str
    invoice_count: int
    total_outstanding: str
    #: 🔴 How many rows were aged from an assumed due date. A report that
    #: silently invents one for every invoice missing it looks authoritative
    #: and is partly fiction.
    assumed_due_dates: int
    buckets: list[AgeingBucket]
    display: dict[str, str]
    note: str


class AgeingReport(BaseModel):
    summary: AgeingSummary
    rows: list[dict[str, Any]]
    by_buyer: list[dict[str, Any]]


class PriorityOut(BaseModel):
    invoice_id: str
    score: int
    band: str
    factors: list[dict[str, Any]]
    disclaimer: str


# ---------------------------------------------------------------------------
# Payments
# ---------------------------------------------------------------------------


class PaymentRequestCreate(BaseModel):
    """
    🔴 Creating this asks for money. It never records any.

    A manual UPI request comes back `awaiting_manual_confirmation` and stays
    there until a person enters a receipt or a signed webhook matches it.
    """

    provider: str = "manual_upi"
    amount: Decimal | None = None
    note: str | None = None
    idempotency_key: str | None = None


class PaymentRequestOut(BaseModel):
    id: uuid.UUID
    invoice_id: uuid.UUID
    provider: str
    provider_reference: str | None
    amount: Decimal
    currency: str
    payload_url: str | None
    qr_svg: str | None
    status: str
    expires_at: str | None
    created_at: str
    #: Spelled out in the payload so a UI cannot accidentally render this as
    #: a completed payment.
    is_payment: bool = False
    note: str = (
        "A payment request is not a payment. Only a human-entered receipt or a "
        "signed, matched gateway webhook creates one."
    )


class PromiseCreate(BaseModel):
    promised_on: date
    amount: Decimal | None = None
    note: str | None = None
    contact_name: str | None = None


class PromiseOut(BaseModel):
    id: uuid.UUID
    invoice_id: uuid.UUID
    promised_on: date
    promised_amount: Decimal | None
    note: str | None
    contact_name: str | None
    created_at: str


# ---------------------------------------------------------------------------
# Delivery
# ---------------------------------------------------------------------------


class DeliveryPreviewRequest(BaseModel):
    channel: str = "email"
    recipient: str | None = None
    subject: str | None = None
    body: str | None = None
    attach_pdf: bool = True


class DeliveryPreviewOut(BaseModel):
    invoice_id: str
    channel: str
    recipient: str
    recipient_name: str | None
    subject: str | None
    body: str
    pdf_sha256: str | None
    template_version: str
    #: 🔴 Quote this back to send. It covers recipient, subject, body and the
    #: PDF hash together.
    preview_sha256: str
    warnings: list[str]
    blocked_reason: str | None
    can_send: bool


class DeliverySendRequest(DeliveryPreviewRequest):
    """Confirm a frozen preview. The hash is required."""

    preview_sha256: str = Field(min_length=64, max_length=64)
    idempotency_key: str | None = None


class DeliveryOut(BaseModel):
    id: uuid.UUID
    invoice_id: uuid.UUID
    channel: str
    recipient: str
    status: str
    attempts: int
    pdf_sha256: str | None
    provider: str | None
    provider_message_id: str | None
    error_code: str | None
    error_detail: str | None


# ---------------------------------------------------------------------------
# Reminders
# ---------------------------------------------------------------------------


class ReminderPolicyCreate(BaseModel):
    billing_entity: uuid.UUID
    name: str = Field(min_length=1, max_length=120)
    channel: str = "email"
    trigger_days: list[int] = [7, 15, 30]
    template_body: str | None = None
    quiet_hour_start: int = 20
    quiet_hour_end: int = 9
    timezone: str = "Asia/Kolkata"
    min_days_between: int = 7
    max_per_invoice: int = 4


class ReminderPolicyOut(BaseModel):
    id: uuid.UUID
    billing_entity: uuid.UUID
    name: str
    channel: str
    trigger_days: list[int]
    template_version: str
    quiet_hour_start: int
    quiet_hour_end: int
    timezone: str
    min_days_between: int
    max_per_invoice: int
    #: 🔴 Off unless a named person turned it on with a per-run ceiling.
    autosend_enabled: bool
    autosend_max_per_run: int
    is_active: bool


class AutosendRequest(BaseModel):
    """
    Enable or disable unattended sending for a policy.

    🔴 Enabling requires a limit above zero and a reason. Both are recorded,
    because "who decided the machine could message customers unattended, and
    what did they think the ceiling was" is the question afterwards.
    """

    enabled: bool
    max_per_run: int = 0
    reason: str = Field(min_length=3, max_length=500)


class ReminderRunOut(BaseModel):
    id: uuid.UUID
    policy_id: uuid.UUID | None
    status: str
    preview_sha256: str
    candidates: list[dict[str, Any]]
    #: 🔴 Returned alongside the candidates, never hidden. The opt-outs and the
    #: capped invoices are the part of a batch worth reviewing.
    skipped: list[dict[str, Any]]
    expires_at: str
    created_at: str


class ReminderConfirm(BaseModel):
    preview_sha256: str = Field(min_length=64, max_length=64)


class ReminderRunResult(BaseModel):
    run: ReminderRunOut
    queued: int
    reminders: list[dict[str, Any]]
